src/icatcher/cli.py

```diff
@@ -262,14 +252,48 @@ def predict_from_video(opt):
         hor, ver = 0.5, 1  # initial guess for face location
         cur_fps = video.FPS()  # for debugging purposes
         last_class_text = ""  # Initialize so that we see the first class assignment as an event to record
+
+        # if going to use cpu parallelization, don't allow for live stream video
+        if use_cpu and opt.fd_model == "retinaface":
+            # figure out how many cpus can be used
+            num_cpus = cpu_count() - opt.num_cpus_saved
+            assert num_cpus > 0
+
+            # send all frames in to be preprocessed and have faces detected prior to running gaze detection
+            total_frames = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))
+            vid_frames = range(0, total_frames, 1 + opt.fd_skip_frames)  # adding step if frames are skipped
+            processed_frames = process_frames(cap, vid_frames, h_start_at, w_start_at, w_end_at)
```

Comment:
This crops the frames instead of masking them like we do below (see "mask_regions").
The behavior here is unexpected, since if you crop, the coordinates of the image are not the same relative to a masked frame which will be used for cropping the bounding box.

To fix I think you should use the mask_regions functions instead of cropping (inside of process frames? up to you.)

Also doesn't this mean we need to buffer the whole video to memory? This should definitely be optional and off by default. How much faster does this make things go?

Reply:
Exactly !
Buffering sounds totally reasonable for getting performance, just shouldn't be the default imo.